src/core/meshcop/border_agent.hpp

```diff
@@ -387,13 +410,116 @@ class BorderAgent : public InstanceLocator, private NonCopyable
 
     static Coap::Message::Code CoapCodeFromError(Error aError);
 
+    class MeshCopTxtEncoder : public InstanceLocator
+    {
+    public:
+        explicit MeshCopTxtEncoder(Instance &aInstance)
```

Comment:
The `MeshCoPTxtEncoder` now only tracks the `Instance` and we have a member variable `mMeshCopTxtEncoder`.

Some suggestions that I think may help simplify the code:

*   We can have the `MeshCoPTxtEncoder` itself hold the buffer (with a maximum size) and have a member variable `mAppender` as well.
*   This way we don't need to pass `aAppender` in the `AppendTxtEntry()` methods (we can use `mAppender`).
*   In `BorderAgent::NotifyMeshCoPServiceChange(void)`, we can create a local `MeshCoPTxtEncoder` object and use it to encode the TXT data to pass in the callback.

What do you think?

Reply:
Good suggestion! I've updated. The code looks much nicer.